### PR TITLE
Feature/Transaction metadata

### PR DESCRIPTION
**Pull request by DeeJayElly**
Feature/Transaction metadata

Screenshots:

Screen Shot 2020-11-22 at 12 36 34 PM

Screen Shot 2020-11-09 at 2 13 00 PM

Screen Shot 2020-11-09 at 2 13 08 PM

Screen Shot 2020-12-10 at 11 29 44 AM

Screen Shot 2020-12-10 at 11 29 35 AM

--

## Test Cases

**Scenario 1: Transaction with metadata, show** 
Given I am on Cardano explorer app
And default language is ‘lang’
And I have a transaction with metadata
When I search for the transaction
Then I will see a dialog titled ’title_text’
And I will see the content ‘content_text’
When I select button 'show_text'
Then the dialog will be dismissed
And Metadata will be shown
And Metadata is formatted in readable manner

Example:
|lang|title_text|content_text|show_text|
|--|--|--|--|
|en|Warning: Unmoderated content|Transaction metadata is not moderated, and it may contain inappropriate content. The creator of the transaction provides this information and is not in control of this website's operator. Do you want to see the unmoderated content? |Show unmoderated content|
|jp|警告：非承認制コンテンツ|トランザクションメタデータは承認制ではなく、不適切な内容を含む場合があります。この情報はトランザクションの作成者が提供するもので、ウェブサイト管理者の管理下にはありません。管理されていないコンテンツを表示しますか。|管理されていないコンテンツを表示する|

**Scenario 2: Transaction with metadata, leave** 
Given I am on Cardano explorer app
And default language is ‘lang’
And I have a transaction with metadata
When I search for the transaction
Then I will see a button 'leave_text'
When I select the button
Then the dialog will be dismissed
And I am back on Cardano App search page

Example:
|lang|leave_text|
|--|--|
|en|Leave this page|
|jp|このページを離れる|

**Scenario 3: Transaction without metadata**
Given I am on Cardano explorer app
And default language is ‘lang’
And I have a transaction without metadata
When I search for the transaction
Then I will not see a dialog titled ‘title_text’

Example:
|lang|title_text|
|--|--|
|en|Warning: Unmoderated content|
|jp|警告：非承認制コンテンツ|

## Testing Summary

 For Hydra build #5131736:

|Test cases | Results|Note|
|-----------|----|----|
|Scenario 1 | Pass||
|Scenario 2| Pass||
|Scenario 3| Pass||
--

### CODE REVIEW

**Comment by DeeJayElly** [comment]
@DominikGuzei @rhyslbw This PR is updated with new updates for Metadata feature. Pls review it.

**Review by DominikGuzei — CHANGES_REQUESTED** [review]
@DeeJayElly the build fails due to a Typescript error: https://app.netlify.com/sites/cardano-explorer-mainnet/deploys/5fbb8804c4daca000821741f

**Comment by DeeJayElly** [comment]
> @DeeJayElly the build fails due to a Typescript error: https://app.netlify.com/sites/cardano-explorer-mainnet/deploys/5fbb8804c4daca000821741f

Fixed @DominikGuzei @rhyslbw

**Review by DominikGuzei — APPROVED** [review]

**Comment by DominikGuzei** [comment]
@rhyslbw please check again - I fixed the issues I have found and looks good now 👍

**Review by rhyslbw — CHANGES_REQUESTED** [review]
@DeeJayElly 

I'm seeing an error: 

Screenshot from 2020-11-30 20-40-00


Some transactions on the `mainnet` with metadata:

- `43c7a108ab69847ec2bdc3e283dca65a882271a41f01b19a008043494ff128be`
- `f910021138e553c65b96cf3e4647927fcd9f634e06544251f83cffb1891876e8`
- `2532ea3bcb91da8b40785aa6864fc65d022757219c7d406837e50daae475f163`

Also, please ensure the warning is only show if the transaction contains metadata

**Comment by DeeJayElly** [comment]
> @DeeJayElly
> 
> I'm seeing an error:
> 
> Screenshot from 2020-11-30 20-40-00
> 
> Some transactions on the `mainnet` with metadata:
> 
> * `43c7a108ab69847ec2bdc3e283dca65a882271a41f01b19a008043494ff128be`
> * `f910021138e553c65b96cf3e4647927fcd9f634e06544251f83cffb1891876e8`
> * `2532ea3bcb91da8b40785aa6864fc65d022757219c7d406837e50daae475f163`
> 
> Also, please ensure the warning is only show if the transaction contains metadata

Fixed

File: source/features/i18n/translations/de.json [hunk 1]
```diff
@@ -161,13 +161,22 @@
...
-    "transactionsLabel": "Transaktionen"
+    "transactionsLabel": "Transaktionen",
+    "unmoderated": "!!!You have consented to view unmoderated transaction data.",
+    "unmoderatedConfirmation": "!!!Yes, show unmoderated content",
+    "unmoderatedContent": "!!!Do you want to see unmoderated content?",
+    "unmoderatedEntitiesControl": "!!!This information is provided and hosted by stake pool operators and it is not in control of entities running this website.",
+    "unmoderatedLeave": "!!!Leave this page",
+    "unmoderatedPoolTickerSymbols": "!!!Transaction metadata is not moderated and may contain inappropriate content.",
+    "unmoderatedTransactionData": "!!!Transaction metadata data is not moderated",
+    "unmoderatedWarning": "!!!Warning:"
```
Comment (DominikGuzei) [inline source/features/i18n/translations/de.json:179]: What about the German translation? Do we still get this from someone?

File: source/features/address/ui/AddressSummary.tsx [hunk 2]
```diff
@@ -70,7 +76,9 @@ const AddressSummary = (props: AddressSummaryProps) => {
                 <div className={styles.infoLabel}>
                   {translate('address.totalWithdrawn')}
                 </div>
-                <div className={styles.infoValue}>{props.totalWithdrawn} ADA</div>
+                <div className={styles.infoValue}>
+                  {props.totalWithdrawn} ADA
+                </div>
```
Comment (DominikGuzei) [inline source/features/address/ui/AddressSummary.tsx:81]: Btw. why are half of the changes only formatting related? This is very confusing for reviews :(

File: source/features/i18n/translations/ja.json [hunk 3]
```diff
@@ -161,13 +161,22 @@
...
-    "transactionsLabel": "トランザクション"
+    "transactionsLabel": "トランザクション",
+    "unmoderated": "!!!You have consented to view unmoderated transaction data.",
+    "unmoderatedConfirmation": "!!!Yes, show unmoderated content",
+    "unmoderatedContent": "!!!Do you want to see unmoderated content?",
+    "unmoderatedEntitiesControl": "!!!This information is provided and hosted by stake pool operators and it is not in control of entities running this website.",
+    "unmoderatedLeave": "!!!Leave this page",
+    "unmoderatedPoolTickerSymbols": "!!!Transaction metadata is not moderated and may contain inappropriate content.",
+    "unmoderatedTransactionData": "!!!Transaction metadata data is not moderated",
+    "unmoderatedWarning": "!!!Warning:"
```
Comment (DominikGuzei) [inline source/features/i18n/translations/ja.json:179]: What about Japanese?

**Review by DominikGuzei — CHANGES_REQUESTED** [review]
@DeeJayElly I have left you some questions - but also can you briefly describe what you changed in your last commits? There have been a lot of formatting changes and it's hard to spot the "real" changes underneath

File: source/features/i18n/translations/de.json (hunk 1 above)
Comment (DeeJayElly) [reply source/features/i18n/translations/de.json]: Added

File: source/features/i18n/translations/ja.json (hunk 3 above)
Comment (DeeJayElly) [reply source/features/i18n/translations/ja.json]: Added

File: source/features/address/ui/AddressSummary.tsx (hunk 2 above)
Comment (DeeJayElly) [reply source/features/address/ui/AddressSummary.tsx:81]: I am always using 'yarn prettier:format' before my commits/push . @DominikGuzei. I assume most of the old changes weren't formatted previously.

**Comment by DeeJayElly** [comment]
> @DeeJayElly I have left you some questions - but also can you briefly describe what you changed in your last commits? There have been a lot of formatting changes and it's hard to spot the "real" changes underneath

@DominikGuzei I've fixed showing of unmoderated data dialog on transactions search page + i've added rendering of metadata in a slightly different way then rest of the rows on the transaction page.

https://github.com/input-output-hk/cardano-explorer-app/pull/372/commits/b738ccf9ed74a9d437da3de33503ffebb45087f4

So because metadata array can contain bunch of different types of data, objects, arrays, strings, etc... we are just formating them as a big JSON string:

`
{props.metadata.map(item => {
              return (
                <div>
                  {JSON.stringify(item.value)}
                </div>
              )
})}
`

File: source/features/transactions/components/TransactionInfo.tsx [hunk 4]
```diff
@@ -256,6 +239,21 @@ const TransactionInfo = (props: ITransactionInfoProps) => {
...
+      {/* ===== METADATA ===== */}
+
+      {props.metadata && props.metadata.length && (
+        <div className={styles.row}>
+          <div className={styles.label}>
+            {translate('transaction.metadata')}
+          </div>
+          <div className={styles.value}>
+            {props.metadata.map((item) => {
+              return <div>{JSON.stringify(item.value)}</div>;
```
Comment (rhyslbw) [inline source/features/transactions/components/TransactionInfo.tsx:251]:
```suggestion
              return <div>{JSON.stringify(item.value, undefined, 2)}</div>;
```

**Review by rhyslbw — CHANGES_REQUESTED** [review]
1. The warning should **only** be shown if the transaction has metadata.
2. The formatting can be improved using my inline suggestion

File: source/features/transactions/components/TransactionInfo.tsx (hunk 4 above)
Comment (DeeJayElly) [reply source/features/transactions/components/TransactionInfo.tsx]: Added

**Comment by DeeJayElly** [comment]
> 1. The warning should **only** be shown if the transaction has metadata.
> 2. The formatting can be improved using my inline suggestion

1. Fixed
2. Added

**Review by rhyslbw — CHANGES_REQUESTED** [review]
Is there a styling override causing the `JSON.stringify` to not present the object with leading white space?

Screenshot from 2020-12-02 00-15-26

**Comment by DeeJayElly** [comment]
> Is there a styling override causing the `JSON.stringify` to not present the object with leading white space?
> 
> Screenshot from 2020-12-02 00-15-26

I inspected it, its just rendering like that all the time.

**Comment by DeeJayElly** [comment]
> > Is there a styling override causing the `JSON.stringify` to not present the object with leading white space?
> > Screenshot from 2020-12-02 00-15-26
> 



> Is there a styling override causing the `JSON.stringify` to not present the object with leading white space?
> 
> Screenshot from 2020-12-02 00-15-26

Fixed @rhyslbw @DominikGuzei 

Screen Shot 2020-12-01 at 8 34 06 PM

**Review by rhyslbw — APPROVED** [review]

**Review by DominikGuzei — CHANGES_REQUESTED** [review]
@DeeJayElly @rhyslbw I talked to Darko and he said this:
"For stake pools, we need it once. But for metadata every time."

So I guess we need to change the logic for transactions to show the warning every time (only for tx that have the metadata) … maybe we should make this also less intrusive? Like, showing the warning only when the user actually wants to see the metadata?

**Review by rhyslbw — CHANGES_REQUESTED** [review]
Screenshot from 2020-12-07 21-55-16

**Comment by DeeJayElly** [comment]
> @DeeJayElly @rhyslbw I talked to Darko and he said this:
> "For stake pools, we need it once. But for metadata every time."
> 
> So I guess we need to change the logic for transactions to show the warning every time (only for tx that have the metadata) … maybe we should make this also less intrusive? Like, showing the warning only when the user actually wants to see the metadata?

@DominikGuzei @rhyslbw I have implemented the requested change. Can you guys do another round of review.

**Review by rhyslbw — CHANGES_REQUESTED** [review]
1. The warning must be shown every time, not once per session
2. Render the component in place of the metadata, so that the other details are shown

Screenshot from 2020-12-09 12-26-05

File: source/features/search/ui/TransactionSearchResult.tsx [hunk 5]
```diff
@@ -37,14 +40,20 @@ export const TransactionSearchResult = () => {
...
           return (
-            <TransactionInfo
-              dontLinkToTransaction
-              navigation={navigation?.actions}
-              networkBlockHeight={networkInfo.store.blockHeight}
-              title={translate('transaction.transactionLabel')}
-              showDetails
-              {...transactionSearchResult}
-            />
+            <>
```
Comment (DominikGuzei) [inline source/features/search/ui/TransactionSearchResult.tsx:43]: i think you can remove the React Fragment wrapper here as the TX info is the only component to render

File: source/features/search/ui/TransactionSearchResult.tsx [hunk 6]
```diff
@@ -17,6 +17,9 @@ export const TransactionSearchResult = () => {
   const search = useSearchFeature();
   const networkInfo = useNetworkInfoFeature();
   const navigation = useNavigationFeature();
+  const [isShowingUnmoderatedData, setisShowingUnmoderatedData] = useState(
+    true
+  );
```
Comment (DominikGuzei) [inline source/features/search/ui/TransactionSearchResult.tsx:22]: Is there a specific reason why the state for moderated data is kept in the search result and not directly in the `TransactionInfo` component?

**Review by DominikGuzei — CHANGES_REQUESTED** [review]

File: source/features/search/ui/TransactionSearchResult.tsx (hunk 5 above)
Comment (DeeJayElly) [reply source/features/search/ui/TransactionSearchResult.tsx:43]: Done

File: source/features/search/ui/TransactionSearchResult.tsx (hunk 6 above)
Comment (DeeJayElly) [reply source/features/search/ui/TransactionSearchResult.tsx]: I're right, overcomplicating without a reason. I have simplified it now.

**Comment by DeeJayElly** [comment]
> 1. The warning must be shown every time, not once per session
> 2. Render the component in place of the metadata, so that the other details are shown
> 
> Screenshot from 2020-12-09 12-26-05

Fixed

**Review by rhyslbw — DISMISSED** [review]
Good result @DeeJayElly  :+1:

**Comment by nikolaglumac** [comment]
@rhyslbw please take care of merging and deploying this change 🙏

**Comment by mioriohk** [comment]
Hi @DeeJayElly. For Hydra build #5131736. Looks good to me. 👍 

|Test cases | Results|Note|
|-----------|----|----|
|Scenario 1 | Pass||
|Scenario 2| Pass||
|Scenario 3| Pass||

**Review by DominikGuzei — APPROVED** [review]

**Pull request merged**